src/storage/aggregator.cc

```diff
@@ -0,0 +1,383 @@
+    : Aggregator(base_meta, aggr_meta, aggr_table, index_pos, aggr_col, aggr_type, ts_col, window_tpye, window_size) {}
+
+bool SumAggregator::UpdateAggrVal(const codec::RowView& row_view, const int8_t* row_ptr, AggrBuffer* aggr_buffer) {
+    if (row_view.IsNULL(row_ptr, aggr_col_idx_)) {
+        return true;
+    }
+    switch (aggr_col_type_) {
+        case DataType::kSmallInt: {
+            int16_t val;
+            row_view.GetValue(row_ptr, aggr_col_idx_, aggr_col_type_, &val);
+            aggr_buffer->aggr_val_.vlong += val;
+            break;
+        }
+        case DataType::kInt: {
+            int32_t val;
+            row_view.GetValue(row_ptr, aggr_col_idx_, aggr_col_type_, &val);
+            aggr_buffer->aggr_val_.vlong += val;
+            break;
+        }
+        case DataType::kBigInt: {
+            int64_t val;
+            row_view.GetValue(row_ptr, aggr_col_idx_, aggr_col_type_, &val);
+            aggr_buffer->aggr_val_.vlong += val;
+            break;
+        }
+        case DataType::kFloat: {
+            float val;
+            row_view.GetValue(row_ptr, aggr_col_idx_, aggr_col_type_, &val);
+            aggr_buffer->aggr_val_.vfloat += val;
+            break;
+        }
+        case DataType::kDouble: {
+            double val;
+            row_view.GetValue(row_ptr, aggr_col_idx_, aggr_col_type_, &val);
+            aggr_buffer->aggr_val_.vdouble += val;
+            break;
+        }
+        default: {
+            PDLOG(ERROR, "Unsupported data type");
+            return false;
+        }
+    }
+    return true;
+}
+
+std::shared_ptr<Aggregator> CreateAggregator(const ::openmldb::api::TableMeta& base_meta,
+                                             const ::openmldb::api::TableMeta& aggr_meta,
+                                             std::shared_ptr<Table> aggr_table, const uint32_t& index_pos,
+                                             const std::string& aggr_col, const std::string& aggr_func,
+                                             const std::string& ts_col, const std::string& bucket_size) {
+    std::string aggr_type = boost::to_lower_copy(aggr_func);
+    WindowType window_type;
+    uint32_t window_size;
+    if (::openmldb::base::IsNumber(bucket_size)) {
+        window_type = WindowType::kRowsNum;
+        window_size = std::stoi(bucket_size);
+    } else {
+        window_type = WindowType::kRowsRange;
+        if (bucket_size.empty()) {
+            PDLOG(ERROR, "Bucket size is empty");
+            return std::shared_ptr<Aggregator>();
+        }
+        char time_unit = bucket_size.back();
+        std::string time_size = bucket_size.substr(0, bucket_size.size() - 1);
+        boost::trim(time_size);
+        if (!::openmldb::base::IsNumber(time_size)) {
+            PDLOG(ERROR, "Bucket size is not a number");
+            return std::shared_ptr<Aggregator>();
+        }
+        switch (time_unit) {
+            case 's':
```

Comment:
we also have "ms" unit.

Reply:
> we also have "ms" unit.

however, when parsing SQL, there is no implementation for `ms`. see `ASTIntervalLIteralToNum`

Reply:
oh. ic